support with plugins block with kotlin

**settings.gradle**
```groovy:settings.gradle
buildscript {
    repositories {
        mavenCentral()
        maven(url = "https://files.minecraftforge.net/maven") {
            name = "forge"
        }
    }
    dependencies {
        classpath("com.anatawa12.forge:ForgeGradle:2.3-1.0.+") {
            isChanging = true
        }
    }
}
```

**build.gradle**
```groovy:build.gradle
plugins {
    kotlin("jvm") version "1.4.20"
    id("net.minecraftforge.gradle.forge")
}
```

then get

```
> Configure project :
Missing FG dep in buildscript classpath. Forking decompilation is likely to break.

FAILURE: Build failed with an exception.

* What went wrong:
org/jetbrains/kotlin/gradle/plugin/KotlinSourceSet
> org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.8.3/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 1s
```

`org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet` is class not found error/exception.

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -119,7 +119,7 @@ dependencies {
        exclude(group = "org.ow2.asm")
     }
 
-    compileOnly("org.jetbrains.kotlin:kotlin-gradle-plugin:1.1.3-2")
+    //compileOnly("org.jetbrains.kotlin:kotlin-gradle-plugin:1.1.3-2")
     testCompile("junit:junit:4.12")
 }
 

--- a/src/main/java/net/minecraftforge/gradle/common/BasePlugin.java
+++ b/src/main/java/net/minecraftforge/gradle/common/BasePlugin.java
@@ -33,6 +33,7 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 
+import net.minecraftforge.gradle.util.ReflectionUtil;
 import net.minecraftforge.gradle.util.json.version.ManifestVersion;
 import org.gradle.api.Action;
 import org.gradle.api.DefaultTask;
@@ -47,6 +48,8 @@ import org.gradle.api.artifacts.dsl.DependencyHandler;
 import org.gradle.api.artifacts.repositories.ArtifactRepository;
 import org.gradle.api.artifacts.repositories.FlatDirectoryArtifactRepository;
 import org.gradle.api.artifacts.repositories.MavenArtifactRepository;
+import org.gradle.api.initialization.Settings;
+import org.gradle.api.initialization.dsl.ScriptHandler;
 import org.gradle.api.logging.Logger;
 import org.gradle.api.logging.Logging;
 import org.gradle.api.plugins.ExtraPropertiesExtension;
@@ -305,6 +308,13 @@ public abstract class BasePlugin<K extends BaseExtension> implements Plugin<Proj
         } catch (IllegalAccessException | NoSuchFieldException e) {
             throw new RuntimeException(e);
         }
+        Settings settings = ReflectionUtil.getSettingsOrNull(project.getGradle());
+        return hasMavenCentralBeforeJCenter(project.getBuildscript(), mavenCentralUrl) ||
+                (settings != null && hasMavenCentralBeforeJCenter(settings.getBuildscript(), mavenCentralUrl));
+    }
+
+    private boolean hasMavenCentralBeforeJCenter(
+            ScriptHandler buildScript, java.net.URI mavenCentralUrl) {
         for (ArtifactRepository repository : project.getBuildscript().getRepositories()) {
             if (repository instanceof MavenArtifactRepository) {
                 MavenArtifactRepository mvnRepo = (MavenArtifactRepository) repository;
@@ -402,15 +412,17 @@ public abstract class BasePlugin<K extends BaseExtension> implements Plugin<Proj
         Project parent = project;
         Dependency fgDepTemp = null;
         Configuration buildscriptClasspath = null;
-        while (parent != null && fgDepTemp == null) {
-            buildscriptClasspath = parent.getBuildscript().getConfigurations().getByName("classpath");
-            fgDepTemp = Iterables.getFirst(buildscriptClasspath.getDependencies().matching(new Spec<Dependency>() {
-                @Override
-                public boolean isSatisfiedBy(Dependency element)
-                {
-                    return element.getName().equals(GROUP_FG);
-                }
-            }), null);
+        while (fgDepTemp == null) {
+            if (parent != null) {
+                buildscriptClasspath = parent.getBuildscript().getConfigurations().getByName("classpath");
+            } else {
+                Settings settings = ReflectionUtil.getSettingsOrNull(project.getGradle());
+                if (settings == null) break;
+                buildscriptClasspath = settings.getBuildscript().getConfigurations().getByName("classpath");
+            }
+            fgDepTemp = Iterables.getFirst(buildscriptClasspath.getDependencies()
+                    .matching(element -> element.getName().equals(GROUP_FG)), null);
+            if (parent == null) break;
             parent = parent.getParent();
         }
         final Dependency fgDep = fgDepTemp;

--- a/src/main/java/net/minecraftforge/gradle/user/UserBasePlugin.java
+++ b/src/main/java/net/minecraftforge/gradle/user/UserBasePlugin.java
@@ -25,6 +25,7 @@ import static net.minecraftforge.gradle.user.UserConstants.*;
 
 import java.io.File;
 import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.HashMap;
@@ -39,6 +40,8 @@ import javax.xml.transform.TransformerFactory;
 import javax.xml.transform.dom.DOMSource;
 import javax.xml.transform.stream.StreamResult;
 
+import net.minecraftforge.gradle.util.ReflectionUtil;
+import org.codehaus.groovy.runtime.callsite.CallSite;
 import org.gradle.api.Action;
 import org.gradle.api.DefaultTask;
 import org.gradle.api.NamedDomainObjectContainer;
@@ -60,6 +63,7 @@ import org.gradle.api.artifacts.result.DependencyResult;
 import org.gradle.api.artifacts.result.ResolvedArtifactResult;
 import org.gradle.api.file.ConfigurableFileCollection;
 import org.gradle.api.file.FileTreeElement;
+import org.gradle.api.file.SourceDirectorySet;
 import org.gradle.api.internal.plugins.DslObject;
 import org.gradle.api.plugins.JavaPlugin;
 import org.gradle.api.plugins.JavaPluginConvention;
@@ -70,6 +74,7 @@ import org.gradle.api.tasks.JavaExec;
 import org.gradle.api.tasks.ScalaSourceSet;
 import org.gradle.api.tasks.SourceSet;
 import org.gradle.api.tasks.bundling.Jar;
+import org.gradle.api.tasks.compile.AbstractCompile;
 import org.gradle.api.tasks.compile.GroovyCompile;
 import org.gradle.api.tasks.compile.JavaCompile;
 import org.gradle.api.tasks.javadoc.Javadoc;
@@ -78,8 +83,6 @@ import org.gradle.jvm.JvmLibrary;
 import org.gradle.language.base.artifact.SourcesArtifact;
 import org.gradle.plugins.ide.eclipse.model.EclipseModel;
 import org.gradle.plugins.ide.idea.model.IdeaModel;
-import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet;
-import org.jetbrains.kotlin.gradle.tasks.KotlinCompile;
 import org.w3c.dom.DOMException;
 import org.w3c.dom.Document;
 import org.w3c.dom.Element;
@@ -609,16 +612,16 @@ public abstract class UserBasePlugin<T extends UserBaseExtension> extends BasePl
                 // kotlin
                 if (project.getPlugins().hasPlugin("kotlin"))
                 {
-                    KotlinSourceSet langSet = (KotlinSourceSet) new DslObject(set).getConvention().getPlugins().get("kotlin");
+                    Object langSet = new DslObject(set).getConvention().getPlugins().get("kotlin");
                     File dir = new File(dirRoot, "kotlin");
 
                     task = makeTask(taskPrefix+"Kotlin", TaskSourceCopy.class);
-                    task.setSource(langSet.getKotlin());
+                    task.setSource((SourceDirectorySet)ReflectionUtil.callMethod(langSet, "getKotlin"));
                     task.setOutput(dir);
 
                     // must get replacements from extension afterEValuate()
 
-                    KotlinCompile compile = (KotlinCompile) project.getTasks().getByName(set.getCompileTaskName("kotlin"));
+                    AbstractCompile compile = (AbstractCompile) project.getTasks().getByName(set.getCompileTaskName("kotlin"));
                     compile.dependsOn(task);
                     compile.setSource(dir);
                     Path dirPath = dir.toPath();
@@ -889,8 +892,12 @@ public abstract class UserBasePlugin<T extends UserBaseExtension> extends BasePl
                 }
                 if (project.getPlugins().hasPlugin("kotlin"))
                 {
-                    KotlinSourceSet langSet = (KotlinSourceSet) new DslObject(main).getConvention().getPlugins().get("kotlin");
-                    sourceJar.from(langSet.getKotlin());
+                    Object langSet = new DslObject(main).getConvention().getPlugins().get("kotlin");
+                    try {
+                        sourceJar.from(langSet.getClass().getMethod("getKotlin").invoke(langSet));
+                    } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+                        throw new RuntimeException(e);
+                    }
                 }
             }
         });

--- a/src/main/java/net/minecraftforge/gradle/util/AnnotationUtils.java
+++ b/src/main/java/net/minecraftforge/gradle/util/AnnotationUtils.java
@@ -25,7 +25,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
 
-import org.jetbrains.org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Opcodes;
 import org.objectweb.asm.AnnotationVisitor;
 import org.objectweb.asm.ClassReader;
 import org.objectweb.asm.ClassVisitor;

--- a/src/main/java/net/minecraftforge/gradle/util/ReflectionUtil.java
+++ b/src/main/java/net/minecraftforge/gradle/util/ReflectionUtil.java
@@ -1,0 +1,41 @@
+/*
+ * A Gradle plugin for the creation of Minecraft mods and MinecraftForge plugins.
+ * Copyright (C) 2013-2019 Minecraft Forge
+ * Copyright (C) 2020-2021 anatawa12 and other contributors
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+package net.minecraftforge.gradle.util;
+
+import java.lang.reflect.InvocationTargetException;
+
+public class ReflectionUtil
+{
+    public static Object callMethod(Object self, String method) {
+        try {
+            return self.getClass().getMethod(method).invoke(self);
+        } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            throw new RuntimeException(e);
+        }
+    }
+    public static Object callMethodOrNull(Object self, String method) {
+        try {
+            return self.getClass().getMethod(method).invoke(self);
+        } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            return null;
+        }
+    }
+}

--- a/src/main/java/net/minecraftforge/gradle/util/ReflectionUtil.java
+++ b/src/main/java/net/minecraftforge/gradle/util/ReflectionUtil.java
@@ -20,6 +20,9 @@
  */
 package net.minecraftforge.gradle.util;
 
+import org.gradle.api.initialization.Settings;
+import org.gradle.api.invocation.Gradle;
+
 import java.lang.reflect.InvocationTargetException;
 
 public class ReflectionUtil
@@ -31,11 +34,17 @@ public class ReflectionUtil
             throw new RuntimeException(e);
         }
     }
-    public static Object callMethodOrNull(Object self, String method) {
+
+    public static <T> T callMethodOrNullWithReturnType(Object self, String method, Class<T> returns) {
         try {
-            return self.getClass().getMethod(method).invoke(self);
-        } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+            return returns.cast(self.getClass().getMethod(method).invoke(self));
+        } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException | ClassCastException e) {
             return null;
         }
+    }
+
+    public static Settings getSettingsOrNull(Gradle gradle) {
+        return ReflectionUtil.callMethodOrNullWithReturnType(gradle,
+                        "getSettings", Settings.class);
     }
 }

